database/sql/V033__new_table_interactions.sql

```diff
@@ -0,0 +1,12 @@
+-- Creates new table for pre_interactions of orders
+
+CREATE TABLE interactions (
+    order_uid bytea NOT NULL,
+    index int NOT NULL,
+    target bytea NOT NULL,
+    value numeric(78,0) NOT NULL,
+    data bytea NOT NULL,
+    PRIMARY KEY(order_uid, index)
+);
+
+CREATE INDEX uid ON interactions USING HASH (order_uid);
```

Comment:
```suggestion
```

Not needed with this primary key.

Reply:
Are you sure? Because we are using it in the searches queries and later on many join operations.

Reply:
An index on (order_uid, index) is more general than just order_uid so it *can* definitely be used in it's place. Whether Postgres decides to do so is another matter but I think it would.

Reply:
I did a small test like this:
```
    #[tokio::test]
    #[ignore]
    async fn postgres_interaction_performance() {
        // the following test can be used a performance test, if the values in the for loop are
        // increased
        let mut db = PgConnection::connect("postgresql://").await.unwrap();
        let mut db = db.begin().await.unwrap();
        crate::clear_DANGER_(&mut db).await.unwrap();

        type Data = ([u8; 56], Address, DateTime<Utc>);
        async fn user_orders(
            ex: &mut PgConnection,
            owner: &Address,
            offset: i64,
            limit: Option<i64>,
        ) -> Vec<Data> {
            super::user_orders(ex, owner, offset, limit)
                .map(|o| {
                    let o = o.unwrap();
                    (o.uid.0, o.owner, o.creation_timestamp)
                })
                .collect::<Vec<_>>()
                .await
        }

        // fill the database
        for i in 0..10000u32 {
            let mut i_as_bytes = i.to_ne_bytes().to_vec();
            let mut order_uid_info = vec![0; 56 - i_as_bytes.len()];
            i_as_bytes.append(&mut order_uid_info);
            let order_uid = ByteArray(i_as_bytes.try_into().unwrap());
            insert_order(
                &mut db,
                &Order {
                    uid: order_uid.clone(),
                    ..Default::default()
                },
            )
            .await
            .unwrap();
            insert_pre_interactions(
                &mut db,
                &vec![(order_uid, Interaction::default()); 10].as_slice(),
            )
            .await
            .unwrap();
        }

        let now = Utc::now();
        single_full_order(&mut db, &ByteArray([0u8; 56]))
            .await
            .unwrap()
            .unwrap();
        let time_diff = Utc::now() - now;
        println!("{:?}", time_diff);
        assert!(time_diff < chrono::Duration::seconds(1));
    }
```
and you are right! Nice.